### PR TITLE
Update Cloudpickle

### PR DESCRIPTION
**Pull request by richardliaw**
## Why are these changes needed?

cc @suquark

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.

### CODE REVIEW

**Comment by robertnishihara** [comment]
One of the main reasons for doing this is that the current version of cloudpickle that we vendor does not properly serialize keyword only arguments in functions.

**Comment by suquark** [comment]
@richardliaw Is it from the master branch of cloudpickle?

**Comment by suquark** [comment]
jenkins, retest this please

**Comment by richardliaw** [comment]
Yeah it is from master

**Comment by pcmoritz** [comment]
Jenkins retest this please

**Comment by suquark** [comment]
Here's one easy patch to the original cloudpickle:

```python
    def save_codeobject(self, obj):
        """
        Save a code object
        """
        if PY3:  # pragma: no branch
            args = (
                obj.co_argcount, obj.co_kwonlyargcount, obj.co_nlocals, obj.co_stacksize,
                obj.co_flags, obj.co_code, obj.co_consts, obj.co_names, obj.co_varnames,
                obj.co_filename, obj.co_name, obj.co_firstlineno, obj.co_lnotab, obj.co_freevars,
                obj.co_cellvars
            )
        else:
            args = (
                obj.co_argcount, obj.co_nlocals, obj.co_stacksize, obj.co_flags, obj.co_code,
                obj.co_consts, obj.co_names, obj.co_varnames, obj.co_filename, obj.co_name,
                obj.co_firstlineno, obj.co_lnotab, obj.co_freevars, obj.co_cellvars
            )
        self.save_reduce(types.CodeType, args, obj=obj)
```

**Comment by richardliaw** [comment]
@suquark we should update the cloudpickle entirely rather than monkey patch it.

File: rllib/tests/test_lstm.py [hunk 1]
```diff
@@ -84,6 +84,8 @@ class RNNSpyModel(Model):
     capture_index = 0
 
     def _build_layers_v2(self, input_dict, num_outputs, options):
+        # This is needed as cloudpickle will reference the same class.
+        RNNSpyModel.capture_index = 0
```
Comment (richardliaw) [inline rllib/tests/test_lstm.py]: cc @ericl

**Review by pcmoritz — APPROVED** [review]

**Comment by pcmoritz** [comment]
@ericl Do you foresee anything else breaking? It's important to merge this so @suquark can move forward with https://github.com/ray-project/ray/pull/5611

File: rllib/tests/test_lstm.py (hunk 1 above)
Comment (robertnishihara) [reply rllib/tests/test_lstm.py]: I don't really understand the comment. Can you clarify if this fix addresses a cloudpickle bug or an RLlib bug and maybe say a bit more about it?

I saw the jenkins test failure and spent a little time looking into it, but didn't track it down successfully.

File: rllib/tests/test_lstm.py [hunk 2]
```diff
@@ -84,6 +84,8 @@ class RNNSpyModel(Model):
     capture_index = 0
 
     def _build_layers_v2(self, input_dict, num_outputs, options):
+        # This is needed as cloudpickle will reference the same class.
```
Comment (richardliaw) [inline rllib/tests/test_lstm.py]:
```suggestion
        # Previously, a new class object was created during
        # deserialization and this `capture_index`
        # variable would be refreshed between class instantiations.
        # This behavior is no longer the case, so we manually refresh
        # the variable.
```

File: rllib/tests/test_lstm.py (hunk 1 above)
Comment (richardliaw) [reply rllib/tests/test_lstm.py]: It's neither a cloudpickle bug or a rllib bug; the behavior for serializing this test object has changed (but is a testing-specific behavior).

Comment (richardliaw) [reply rllib/tests/test_lstm.py]: updated comment

File: rllib/tests/test_lstm.py [hunk 3]
```diff
@@ -84,6 +84,12 @@ class RNNSpyModel(Model):
     capture_index = 0
 
     def _build_layers_v2(self, input_dict, num_outputs, options):
+        # Previously, a new class object was created during
+        # deserialization and this `capture_index`
+        # variable would be refreshed between class instantiations.
+        # This behavior is no longer the case, so we manually refresh
+        # the variable.
+        RNNSpyModel.capture_index = 0
```
Comment (richardliaw) [inline rllib/tests/test_lstm.py]:
```suggestion
        RNNSpyModel.capture_index = 0

```

**Comment by robertnishihara** [comment]
Thanks @richardliaw, I think we should merge it and try it out. This behavior of reusing existing classes instead of creating a new one seems dangerous to me (e.g., this means that mutations to the class will affect the deserialized object, right?).

@suquark doesn't cloudpickle have an option to turn off this behavior?

**Comment by richardliaw** [comment]
jenkins retest this please

**Review by robertnishihara — APPROVED** [review]

**Comment by robertnishihara** [comment]
@richardliaw just to clarify, you are saying that cloudpickle's behavior has changed between the previous version and this one, right?

**Comment by richardliaw** [comment]
Yeah

**Comment by suquark** [comment]
@robertnishihara I don't think cloudpickle has such options. This is essential for cloudpickle to work correctly with dynamic class between remote processes.

For example,

```python
def main():
    def DynamicClass:
        pass

    @ray.remote
    def remote_function():
        return DynamicClass()

     # register a serializer (assume broadcast to all workers)
     cloudpickle.dispatch[DynamicClass] = ....

     ray.get(remote_function.remote())
```

In this case, if cloudpickle doesn't keep the existing dynamic class, then when it serializes the `DynamicClass()` object (the return value of the remote function), it will try to look up the dispatch table for the correct serializer. However, the returned object's class has been changed, because after `DynamicClass` itself was serialized and deserialized, Python will not think it is the original type (because types are compared by their pointers), even in this case their qualname and content matches. With a different type, the correct serializer will not likely to be triggered. So cloudpickle creates a `class_tracker_id` for each dynamic class, and will reuse the class for instances of dynamic classes.

Pyarrow also did in the same ways but with `class_id`. So the risk could be low.

**Pull request merged**